src/renderer/services/api/BeaconAPIClientMetrics.ts

```diff
@@ -0,0 +1,144 @@
+import {BeaconAPIClient} from "./BeaconAPIClient";
+import {measureExecution} from "../metrics/Metrics";
+import {
+    uint64,
+    BLSPubkey,
+    Epoch,
+    ValidatorDuty,
+    Slot,
+    BeaconBlock,
+    IndexedAttestation,
+    Shard,
+    uint8
+} from "@chainsafe/eth2.0-types";
+import {IBeaconApiClientOptions, ForkInformation, Syncing} from "./interface";
+import {Bucket} from "../db/schema";
+
+export class BeaconAPIClientMetrics extends BeaconAPIClient {
+
+    public constructor(options: IBeaconApiClientOptions) {
```

Comment:
you should receive BeaconAPIClient instance and db instance

Reply:
I tried to create a _plug & play_ metrics service that would allow us to quickly wrap a class and extend it with metrics functionalities.
At first, I didn't want to create a `BeaconAPIClient` implementation that would include metrics as we would need to manually update every occurrence of the "old" `BeaconAPIClient` and as we are adding dependencies to the client (`BeaconAPIClient` instance and `DB` instance) we would need to inject them manually.

I tried to implement metrics in such a way that would allow us to add metrics functionality without any hassle around additional dependencies, that's why I'm extending existing implementations (wrapping existing implementation) and not implementing custom implementation. That's why I'm currently using a "global" DB instance (`getDB()`) and whoever is going to use this client doesn't have to bother with injecting additional dependencies.

Of course, this all makes sense if we are going to use this metrics service later on for other classes, but if we are going to use it only for the `BeaconAPIClient`  we could definitely do it as you suggested (creating a custom implementation by wrapping existing instance of the client and not extending one).

Reply:
we try to avoid singleton database as it makes testing and controlling of database wierd, especially since we operate in two threads so  by all means all dependencies need to be injected in constructor so they can be easily stubed in tests and swapped.

metrics function can be typed so it has to be binded by function which has property db. But to keep it simple for now, allow it to only be binded by BeaconApiClientMetrics instance.

This should allow you to write real unit tests and not worry about api or database implementation like you do now.